aexpect/remote.py

```diff
@@ -718,8 +720,142 @@ def scp_between_remotes(src, dst, port, s_passwd, d_passwd, s_name, d_name,
+    """
+    Copy files to a remote host (guest) through rsync.
+
+    :param host: Hostname or IP address
+    :param port: Port
+    :param username: Username (if required)
+    :param password: Password (if required)
+    :param local_path: Path on the local machine where we are copying from
+    :param remote_path: Path on the remote machine where we are copying to
+    :param limit: Speed limit of file transfer.
+    :param log_filename: If specified, log all output to this file
+    :param log_function: If specified, log all output using this function
+    :param timeout: The time duration (in seconds) to wait for the transfer
+                    to complete.
+    :param interface: The interface the neighbours attach to (only use when using
+                      ipv6 linklocal address).
+    :param directory: True to copy recursively if the directory to rsync
+    :raise: Whatever remote_rsync() raises
+    """
+    if limit:
+        limit = f"--bwlimit={limit}"
+
+    if host and host.lower().startswith("fe80"):
+        if not interface:
+            raise RsyncError("When using ipv6 linklocal address must assign",
+                             "the interface the neighbour attache")
+        host = f"{host}%{interface}"
+
+    command = "rsync"
+    if directory:
+        command = f"{command} -r"
+    command += (f" -avz -e 'ssh -p {port} -o UserKnownHostsFile=/dev/null "
+                f"-o StrictHostKeyChecking=no' {limit} "
+                f"{quote_path(local_path)} {username}@{host}:{shlex.quote(remote_path)}")
+    password_list = [password]
+    return remote_copy(command, password_list,
+                       log_filename, log_function, timeout)
+
+
+def rsync_from_remote(host, port, username, password, remote_path, local_path,
+                      limit="", log_filename=None, log_function=None,
+                      timeout=600, interface=None, directory=True):
+    """
+    Copy files from a remote host (guest) through rsync.
+
+    :param host: Hostname or IP address
+    :param port: Port
+    :param username: Username (if required)
+    :param password: Password (if required)
+    :param local_path: Path on the local machine where we are copying from
+    :param remote_path: Path on the remote machine where we are copying to
+    :param limit: Speed limit of file transfer.
+    :param log_filename: If specified, log all output to this file
+    :param log_function: If specified, log all output using this function
+    :param timeout: The time duration (in seconds) to wait for the transfer
+                    to complete.
+    :param interface: The interface the neighbours attach to (only use when
+                      using ipv6 linklocal address).
+    :param directory: True to copy recursively if the directory to rsync
+    :raise: Whatever remote_rsync() raises
+    """
+    if limit:
+        limit = f"--bwlimit={limit}"
+    if host and host.lower().startswith("fe80"):
```

Comment:
This is repeated several-times, how about adding a function for that?

Reply:
I would say let's see if IPv6 really becomes properly supported in the future which would prompt some refactoring of such specialized validation.